File: javascript/153-Find-Minimum-in-Rotated-Sorted-Array.js
/**
 * @param {number[]} nums
 * Time O(log(N)) | Space O(1)
 * @return {number}
 */
 var findMin = function(nums) {
    let [ left, right ] = [ 0, (nums.length - 1) ];

    while (left < right) {
        const mid = (left + right) >> 1;
        const guess = nums[mid];
        const [ leftNum, rightNum ] = [ nums[left], nums[right] ];

        const isTarget = leftNum < rightNum;
        if (isTarget) return leftNum;

        const isTargetGreater = leftNum <= guess;
        if (isTargetGreater) left = mid + 1;

        const isTargetLess = guess < leftNum;
        if (isTargetLess) right = mid;
    }

    return nums[left];
};

File: go/binarySearch/153_find-minimum-in-rotated-sorted-array/153_find-minimum-in-rotated-sorted-array.go
package findminimuminrotatedsortedarray

func findMin(nums []int) int {
	result := nums[0]
	left, right := 0, len(nums)-1
	for left <= right {
		if nums[left] <= nums[right] {
			result = min(result, nums[left])
			break
		}
		mid := (left + right) / 2
		result = min(result, nums[mid])
		if nums[mid] >= nums[left] {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}
	return result
}

func min(i, j int) int {
	if i < j {
		return i
	}
	return j
}
